### PR TITLE
[Qt] Bump version to v5.15.1

### PR DESCRIPTION
**Pull request by barche**
Draft PR to try and add more platforms (notably aarch64).

A local check indicated that at least armv7 doesn't compile anymore, either.

### CODE REVIEW

**Comment by giordano** [comment]
Something is forcing `-march=...`

**Comment by barche** [comment]
> Something is forcing `-march=...`

Thanks, was still piecing together the errors to ask help, but you beat me to it :)

**Comment by barche** [comment]
It seems the armv7 build passed `-march=armv8-a`. Isn't that the one for aarch64? If so, building for aarch64 may be as easy as selecting the raspberry pi target and add it to the platform list?

**Comment by giordano** [comment]
That might be possible, yes

File: Q/Qt/build_tarballs.jl [hunk 1]
```diff
@@ -3,18 +3,21 @@
 using BinaryBuilder
 
 name = "Qt"
-version = v"5.15.0"
+version = v"5.15.1"
 
 # Collection of sources required to build qt5
 sources = [
     ArchiveSource("https://download.qt.io/official_releases/qt/5.15/$version/single/qt-everywhere-src-$version.tar.xz",
```
Comment (giordano) [inline Q/Qt/build_tarballs.jl:10]: You can fully parametrise the URL with
```suggestion
    ArchiveSource("https://download.qt.io/official_releases/qt/$(version.major)$(version.minor)/$version/single/qt-everywhere-src-$version.tar.xz",
```

File: Q/Qt/build_tarballs.jl [hunk 2]
```diff
@@ -95,8 +98,23 @@ EOT
...
+            -skip qtwinextras -fontconfig -sysroot /opt/$target/bin/../$target/sys-root
+        ;;
+        
+    *aarch64-linux*)
+        apk add g++ linux-headers
+        export PATH=$(echo "$PATH" | sed -e 's!/opt/bin:!!')
+        export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$prefix/share/pkgconfig
+        export PKG_CONFIG_LIBDIR=$prefix/lib/pkgconfig
+
+        sed -i 's/^QMAKE_C.*FLAGS.*//' ../qt-everywhere-src-*/qtbase/mkspecs/devices/linux-rasp-pi4-v3d-g++/qmake.conf
```
Comment (giordano) [inline Q/Qt/build_tarballs.jl:115]: This file starts with
```
# qmake configuration for the Raspberry Pi 4 (32-bit) using the Mesa V3D
# graphics stack. (not the Broadcom stack)
```
doesn't sound right

Comment (giordano) [reply Q/Qt/build_tarballs.jl:115]: But the configuration doesn't look too wrong (apart from the hard-float thing which I'm not sure about)

Comment (giordano) [reply Q/Qt/build_tarballs.jl:115]: Can't you just use the generic linux configuration?

**Comment by ViralBShah** [comment]
> It seems the armv7 build passed `-march=armv8-a`. Isn't that the one for aarch64? If so, building for aarch64 may be as easy as selecting the raspberry pi target and add it to the platform list?

That is definitely the wrong architecture flag for `armv7`.

File: Q/Qt/build_tarballs.jl (hunk 2 above)
Comment (ViralBShah) [reply Q/Qt/build_tarballs.jl:115]: Yes, the hard float thing is usually on armv7 and not on aarch64.

Comment (ViralBShah) [reply Q/Qt/build_tarballs.jl:115]: Basically even though raspeberry pi 4 has an armv8 processor, it is a 32-bit OS, and you have to build everything as 32-bit- as if it is armv7 (but presumably some of the armv7 compiler flags don't work).

File: Q/Qt/build_tarballs.jl [hunk 3]
```diff
@@ -95,8 +98,23 @@ EOT
...
         
         ../qt-everywhere-src-*/configure QMAKE_LFLAGS=-liconv -platform linux-g++ -device linux-rasp-pi3-g++ -device-option CROSS_COMPILE=/opt/bin/$target- \
+            -extprefix $prefix $commonoptions \
+            -skip qtwinextras -fontconfig -sysroot /opt/$target/bin/../$target/sys-root
+        ;;
+        
+    *aarch64-linux*)
+        apk add g++ linux-headers
+        export PATH=$(echo "$PATH" | sed -e 's!/opt/bin:!!')
+        export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:$prefix/share/pkgconfig
```
Comment (giordano) [inline Q/Qt/build_tarballs.jl:112]: Why? `${prefix}/share/pkgconfig` is already in `${PKG_CONFIG_PATH}`, see https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/16223e035a1b316f3d6e15a2a63733bd36f3935f/src/Runner.jl#L912

Comment (barche) [reply Q/Qt/build_tarballs.jl]: It gave an error in the past, seems resolved now.

File: Q/Qt/build_tarballs.jl (hunk 2 above)
Comment (barche) [reply Q/Qt/build_tarballs.jl]: Yep, switched to the generic version now, thanks for the suggestions!

File: Q/Qt/build_tarballs.jl (hunk 1 above)
Comment (barche) [reply Q/Qt/build_tarballs.jl]: Done.

**Comment by giordano** [comment]
Maybe add also 32-bit Linux and Windows to reach parity with [supported GR platforms](https://gr-framework.org/c.html#installation)?

**Comment by barche** [comment]
For PPC, the error in config.log is:

```
> powerpc64le-linux-gnu-g++ -c -pipe --sysroot=/opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-gnu/sys-root --sysroot=/opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-gnu/sys-root -O2 -w -fPI
C  -I. -I/workspace/srcdir/qt-everywhere-src-5.15.1/qtbase/mkspecs/linux-ppc64-bb -o main.o main.cpp
> powerpc64le-linux-gnu-g++ -liconv -Wl,-O1 -o opengl main.o   -lGL   
> /opt/powerpc64le-linux-gnu/bin/../lib/gcc/powerpc64le-linux-gnu/7.1.0/../../../../powerpc64le-linux-gnu/bin/ld: warning: libdl.so.2, needed by /opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-g
nu/sys-root/usr/local/lib/libGL.so, not found (try using -rpath or -rpath-link)
> /opt/powerpc64le-linux-gnu/bin/../lib/gcc/powerpc64le-linux-gnu/7.1.0/../../../../powerpc64le-linux-gnu/bin/ld: warning: libpthread.so.0, needed by /opt/powerpc64le-linux-gnu/bin/../powerpc64le-li
nux-gnu/sys-root/usr/local/lib/libGL.so, not found (try using -rpath or -rpath-link)
> /opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-gnu/sys-root/usr/local/lib/libGLX.so.0: undefined reference to `dlclose@GLIBC_2.17'
> /opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-gnu/sys-root/usr/local/lib/libGLX.so.0: undefined reference to `dlopen@GLIBC_2.17'
> /opt/powerpc64le-linux-gnu/bin/../powerpc64le-linux-gnu/sys-root/usr/local/lib/libGLX.so.0: undefined reference to `dlsym@GLIBC_2.17'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:67: opengl] Error 1
```

Any ideas? Could be a problem with the Libglvnd package?

**Comment by giordano** [comment]
You probably need to explicitly pass `-ldl`

**Comment by giordano** [comment]
Or better `-Wl,-rpath-link,${prefix}/lib64`, we have to do it some times for PowerPC, see for example https://github.com/JuliaPackaging/Yggdrasil/blob/6964eb12830d356f3359132889490bec1704c8e1/P/Pango/build_tarballs.jl#L22-L24

**Comment by barche** [comment]
> Or better `-Wl,-rpath-link,/opt/${target}/${target}/sys-root/lib64`

Awesome, that worked locally at least :)

File: Q/Qt/build_tarballs.jl [hunk 4]
```diff
@@ -81,22 +90,50 @@ EOT
...
+        ../qt-everywhere-src-*/configure -platform linux-g++ -xplatform freebsd-g++ -device-option CROSS_COMPILE=/opt/bin/$target- \
+            -extprefix $prefix $commonoptions \
+            -skip qtwinextras -fontconfig -sysroot /opt/$target/bin/../$target/sys-root
+		;;
+    
+    *powerpc64le-linux*)
+        cp -a ../qt-everywhere-src-*/qtbase/mkspecs/linux-aarch64-gnu-g++ $qtsrcdir/qtbase/mkspecs/linux-ppc64-bb
+        sed -i 's/aarch64-/powerpc64le-/g' ../qt-everywhere-src-*/qtbase/mkspecs/linux-ppc64-bb/qmake.conf
+
+        ../qt-everywhere-src-*/configure QMAKE_LFLAGS="-liconv -Wl,-rpath-link,/opt/${target}/${target}/sys-root/lib64" -platform linux-g++ -xplatform linux-ppc64-bb -device-option CROSS_COMPILE=/opt/bin/$target- \
```
Comment (giordano) [inline Q/Qt/build_tarballs.jl:136]: Not sure what happens here.  Are `LFLAGS` flags passed directly to the linker?  `-Wl,...` are flags for the compiler, which will forward them to the linker (the `l` in `-Wl` stands for "linker").  The option for the linker would be `-rpath-link=/opt/${target}/${target}/sys-root/lib64`

Comment (barche) [reply Q/Qt/build_tarballs.jl:140]: I don't know how they are passed exactly, I see it going into `LFLAGS` in the code, but this syntax is at lease used elsewhere: https://doc.qt.io/qt-5.12/qtsensors-grue-import-import-pro.html

Comment (giordano) [reply Q/Qt/build_tarballs.jl:140]: Ok, well, if they suggest to use `-Wl` in the documentation then it should be fine.

**Comment by barche** [comment]
It's green ✅! For me this is OK, I suppose I should squash it into one commit?

**Comment by ViralBShah** [comment]
Nice to have all the architectures for QT.

**Comment by vchuravy** [comment]
@barche thanks for doing this! In particular adding PPC ;)

**Pull request merged**